integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/massindexing/MassIndexingInheritanceIT.java

```diff
@@ -292,6 +399,289 @@ void multipleSameBranch() {
+				.anyMatch( "select count\\([a-z0-9_.]+\\) from root [a-z0-9_.]+ "
+						+ "left join aroot [a-z0-9_.]+ on [a-z0-9_.=]+ "
+						+ "left join broot [a-z0-9_.]+ on [a-z0-9_.=]+ "
+						+ "where .+"
+				)
+				// select a1_0.id from aroot a1_0
+				// select r1_0.id from root r1_0 left join aroot r1_1 on r1_0.id=r1_1.id left join broot r1_2 on r1_0.id=r1_2.id where r1_0.type in (?)
+				.anyMatch( "select [a-z0-9_.]+ from aroot [a-z0-9_.]+" )
+				.anyMatch( "select [a-z0-9_.]+ from root [a-z0-9_.]+ "
+						+ "left join aroot [a-z0-9_.]+ on [a-z0-9_.=]+ "
+						+ "left join broot [a-z0-9_.]+ on [a-z0-9_.=]+ "
+						+ "where .+"
+				)
+				// select a1_0.id from aroot a1_0 where a1_0.id=?
+				// select r1_0.id,r1_0.type from root r1_0 where r1_0.id=?
+				.anyMatch( "select [a-z0-9_.,]+ from aroot [a-z0-9_.]+" )
+				.anyMatch( "select [a-z0-9_.,]+ from root [a-z0-9_.]+ "
+						+ "where .+"
+				);
+
+		backendMock.verifyExpectationsMet();
+	}
+
+	@Test
+	void tablePerClass() {
+		with( sessionFactory ).runNoTransaction( session -> {
+			SearchSession searchSession = Search.session( session );
+			MassIndexer indexer = searchSession.massIndexer( RootTablePerClass.class );
+
+			// add operations on indexes can follow any random order,
+			// since they are executed by different threads
+			backendMock.expectWorks(
+					RootTablePerClass.INDEX, DocumentCommitStrategy.NONE, DocumentRefreshStrategy.NONE )
+					.add( "3000000", b -> b.field( "id", 3000000L ) );
+			backendMock.expectWorks(
+					ARootTablePerClass.INDEX, DocumentCommitStrategy.NONE, DocumentRefreshStrategy.NONE )
+					.add( "3000001", b -> b.field( "id", 3000001L ) );
+			backendMock.expectWorks(
+					BRootTablePerClass.INDEX, DocumentCommitStrategy.NONE, DocumentRefreshStrategy.NONE )
+					.add( "3000002", b -> b.field( "id", 3000002L ) );
+
+
+			// purgeAtStart and mergeSegmentsAfterPurge are enabled by default,
+			// so we expect 1 purge, 1 mergeSegments and 1 flush calls in this order:
+			backendMock.expectIndexScaleWorks( RootTablePerClass.INDEX )
+					.purge()
+					.mergeSegments()
+					.flush()
+					.refresh();
+			backendMock.expectIndexScaleWorks( ARootTablePerClass.INDEX )
+					.purge()
+					.mergeSegments()
+					.flush()
+					.refresh();
+			backendMock.expectIndexScaleWorks( BRootTablePerClass.INDEX )
+					.purge()
+					.mergeSegments()
+					.flush()
+					.refresh();
+
+			try {
+				indexer.startAndWait();
+			}
+			catch (InterruptedException e) {
+				fail( "Unexpected InterruptedException: " + e.getMessage() );
+			}
+
+		} );
+
+		statementInspector.hasSelects( 9 )
+				// select count(atpc1_0.id) from aroottableperclass atpc1_0
+				// select count(btpc1_0.id) from broottableperclass btpc1_0
+				// select count(rtpc1_0.id) from (select id, 0 as clazz_ from roottableperclass union all select id, 1 as clazz_ from aroottableperclass union all select id, 2 as clazz_ from broottableperclass) rtpc1_0 where rtpc1_0.clazz_ in (?)
```

Comment:
this looks strage, with a table per class inheritance 🤔 but ... IDK I'm probably missing something 🙈

Reply:
What looks strange to me is that we (probably?) use `where type in` in the HQL query to begin with... ? Aren't we supposed to detect that we target all subtypes and thus that the `where` is unnecessary?

The generated query doesn't look too odd if you consider that Hibernate ORM can't optimize `where type in (something)` when translating from HQL to SQL if `something` is a parameter. I'm not sure such optimization is implemented anyway.

Reply:
hmm yeah ... 
> TABLE_PER_CLASS => no grouping

sooo if we do not group types in table-per-class - then we aren't targeting all types, aaaand if we aren't targeting all types then ... Since 
```
select e.id from RootTablePerClass
```
should return ids of `RootTablePerClass` and its subclasses ... we can't do that since we aren't grouping subclasses ...so we don't want subclasses and we need a type filter...
So while a simple 
```sql
select id from roottableperclass
```
would've been enough I'm not sure we can do it 🙈 (if that makes sense)

Reply:
Ah, right, so this is the query for the root class, because it's not abstract.

I'd say having a non-abstract root class is rather unusual, or at least is generally just an oversight.

Anyway, for table-per-class, we have a choice between:

* Grouping, and getting unions without the `where _clazz in (...)` all over the place.
* Not grouping, and getting unions with the `where _clazz in (...)` but only for the root class and only if it's concrete.

Since in this case the `where _class in` relies on a "synthetic" column (no idea what's the proper name) that isn't backed by actual data on disk, I think the database is likely to optimize it, so we can probably ignore it.

So... I'd say not grouping is the better of two evils?

Reply:
Note we might be able to get rid of `where _clazz in (...)` if Hibernate ORM gave us a way to tell it "target this type only but not its subtypes" in HQL, without relying on a `where` clause.
I don't think that's currently possible though.